sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -442,7 +446,11 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
       case BuildRight => input ++ buildVars
     }
 
-    if (keyIsUnique) {
+    if (isEmptyHashedRelation) {
+      s"""
+         |// If HashedRelation is empty, hash inner join simply returns nothing.
```

Comment:
This isn't a short-cut and we still need to consume the entire stream side. We need something similar to https://github.com/apache/spark/pull/29389

Reply:
@cloud-fan - yes sorry about it. After checking codegen code for example query I can confirm this. For non-codegen (iterator mode) it works, but for codegen it does not work because we are processing in `doConsume()` here so we are still executing the stream side.

So I think
* for non-codegen code path: will keep the same change as in this PR now.
* for codegen code path: do not make change here in `HashJoin`, but adding an adaptive execution logical plan rule e.g. called `EliminateEmptyBroadcastHashJoin.scala` which checks `stage: BroadcastQueryStageExec` to be empty or not (`stage.broadcast.relationFuture.get().value == EmptyHashedRelation`), if it is, then changing the logical plan from `Join`  to `LocalRelation(data = Seq.empty, ...)`.

Does it sound good as plan? Thanks.

Reply:
We can keep this change, as it at least avoids looking up an empty hash relation and is better than before.

We just need to add a new rule, or extend rule `EliminateNullAwareAntiJoin` and rename it to `OptimizeJoinToEmptyRelation`.

Reply:
@cloud-fan - sure, updated.